tests/test_builders/test_build_linkcheck.py

```diff
@@ -1439,3 +1439,72 @@ def test_linkcheck_exclude_documents(app: SphinxTestApp) -> None:
         'uri': 'https://www.sphinx-doc.org/this-is-another-broken-link',
         'info': 'br0ken_link matched br[0-9]ken_link from linkcheck_exclude_documents',
     } in content
+
+
+class CapitalisePathHandler(BaseHTTPRequestHandler):
+    """Test server that uppercases URL paths via redirects."""
+
+    protocol_version = 'HTTP/1.1'
+
+    def do_GET(self):
+        if self.path.startswith('/') and len(self.path) > 1 and self.path[1:].islower():
+            # Redirect lowercase paths to uppercase versions
+            self.send_response(301, 'Moved Permanently')
+            self.send_header('Location', self.path.upper())
+            self.send_header('Content-Length', '0')
+            self.end_headers()
+        elif (
+            self.path.startswith('/') and len(self.path) > 1 and self.path[1:].isupper()
+        ):
+            # Serve uppercase paths
+            content = b'ok\n\n'
+            self.send_response(200, 'OK')
+            self.send_header('Content-Length', str(len(content)))
+            self.end_headers()
+            self.wfile.write(content)
+        else:
+            self.send_response(404, 'Not Found')
+            self.send_header('Content-Length', '0')
+            self.end_headers()
```

Comment:
We can simplify the `if` condition here for future readers, and reduce the number of conditional branches from three to two:

```suggestion
    def do_GET(self):
        if self.path.islower():
            # Redirect lowercase paths to uppercase versions
            self.send_response(301, 'Moved Permanently')
            self.send_header('Location', self.path.upper())
            self.send_header('Content-Length', '0')
            self.end_headers()
        else:
            # Serve uppercase paths
            content = b'ok\n\n'
            self.send_response(200, 'OK')
            self.send_header('Content-Length', str(len(content)))
            self.end_headers()
            self.wfile.write(content)
```